tensorboard/java/org/tensorflow/tensorboard/vulcanize/Vulcanize.java

```diff
@@ -778,6 +783,125 @@ private static String extractAndTransformJavaScript(Document document, Webpath j
     return scriptContent;
   }
 
+  private static void cloneChildrenWithoutWhitespace(Element src, Element dest) {
+    List<Node> toMove = new ArrayList<Node>();
+    for (Node node : src.childNodes()) {
+      if (node instanceof TextNode && ((TextNode) node).isBlank()) {
+        continue;
+      }
+      toMove.add(node);
+    }
+    for (int i = 0; i < toMove.size(); i++) {
+      Node node = toMove.get(i);
+      dest.appendChild(node.clone());
+    }
+  }
+
+  /**
+   * When we inline the HTML based on `<link rel="import">` in `transform`, we
+   * replace the link element with parsed document. This makes us have nested
+   * documents and jsoup's Node.outerHtml (or Node.toString) are incapable of
+   * properly outputting that. Here, we flatten the document by combining all
+   * elements in `<head>` and `<body>` of nested document in one `<head>` and
+   * `<body>`.
+   *
+   * It also prepends <!doctype html> since TensorBoard requires that the
+   * document is HTML.
+   *
+   * NOTE: it makes side-effect to the input `document`.
+   *
+   * Examples:
+   * // Input
+   * <#root> <!-- document -->
+   *   <html>
+   *     <head>
+   *      <#root>
+   *        <html>
+   *          <head>
+   *            <script></script>
+   *            <#root><html><body>welcome </body></html></#root>
+   *          </head>
+   *          <body>foo</body></html>
+   *      </#root></head>
+   *     <body><span>bar</span></body>
+   *   </html>
+   * </html>
+   * // Output
+   * <#root> <!-- document -->
+   *   <!doctype html>
+   *   <html>
+   *     <head><script></script></head>
+   *     <body>welcome foo<span>bar</span></body>
+   *   </html>
+   * </html>
+   **/
+  private static Document getFlattenedHTML5Document(Document document) {
+    Document flatDoc = new Document("/");
+
+    flatDoc.appendChild(new DocumentType("html", "", "", ""));
+
+    // Transfer comment nodes from the `document` level. They are important
+    // license comments
+    for (Node node : document.childNodes()) {
+      if (node instanceof Comment) {
+        flatDoc.appendChild(node.clone());
+      }
+    }
+
+    // Create `<html>`, `<head>` and `<body>`.
+    flatDoc.normalise();
+
+    document.traverse(new FlatDocumentCopier(flatDoc));
+
+    for (Element subdoc : flatDoc.getElementsByTag("#root")) {
+      if (subdoc != flatDoc) {
+        final int MAX_ELEMENT_STR_LEN = 200;
+        String elementStr = subdoc.toString();
+        if (elementStr.length() > MAX_ELEMENT_STR_LEN) {
+          elementStr = elementStr.substring(0, MAX_ELEMENT_STR_LEN) + "...";
+        }
```

Comment:
Not sure how useful it would be to see the document as stringified. Perhaps do the `outerHtml()` or `subdoc.parent().outerHTML()`?

Reply:
Good point. Done.